mixedversion: update skip upgrades logic

RaduBerinde opened the pull request:
This commit updates the logic that determines which versions support
skipping the previous major release during upgrade.

Epic: REL-1292
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134353)
<!-- Reviewable:end -->

RaduBerinde committed:
mixedversion: update skip upgrades logic

This commit updates the logic that determines which versions support
skipping the previous major release during upgrade.

Epic: REL-1292
Release note: None

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion.go:
```diff
@@ -848,26 +869,18 @@ func (t *Test) choosePreviousReleases() ([]*clusterupgrade.Version, error) {
-			return []*clusterupgrade.Version{pred, predPred}, nil
+		// If we haven't performed a skip-version upgrade yet, do it. This logic
+		// makes sure that, when skip-version upgrades are enabled, it happens
+		// when upgrading to the current release, which is the most important
```
> it happens when upgrading to the current release

I think this will also make it so if skip-versions aren't supported to the current release, it will always attempt a skip-version upgrade for the predecessor. I think that's arguably desirable behavior though? (just trying to reason that it's fine)

RaduBerinde replied:
Yeah. I think it's fine because we also have the `skipUpgrades` randomization, so in some runs we won't skip anything.

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mixedversion_test.go:
```diff
@@ -368,3 +369,23 @@ func withTestBuildVersion(v string) func() {
+		if r := clusterversion.Latest.ReleaseSeries(); int(r.Major) == v.Major() && int(r.Minor) == v.Minor() {
+			// We have to special case the current series, to allow for bumping the
+			// min supported version separately from the current version.
+			expected = len(clusterversion.SupportedPreviousReleases()) > 1
```
>to allow for bumping the min supported version separately from the current version
			
Why do we need to allow this? Is it because when we go to `25.1`, we need to bump `minVersion` twice but `currentVersion` once? Or something else?

RaduBerinde replied:
So right now master is still 24.3 and we are first bumping the min version to 24.2, then we will bump the current version to 25.1, then we will bump the min supported again (to 24.3).  We want to separate these because they tend to be large PRs. Ideally we would bump the version to 25.1 first but that cannot be done until we have an RC build for 24.3 (it's needed as a previous release for upgrade testing).

I will consider changing the runbooks to reorder these steps to avoid this issue, in which case I'll clean this up.

DarrylWong replied:
Ack, got it. Thanks for explaining 👍

DarrylWong reviewed (approved):
One small educational question but LGTM besides that.

RaduBerinde commented:
Thank you for the quick review!!

bors r+

srosenberg commented:
> Epic: [REL-1292](https://cockroachlabs.atlassian.net/browse/REL-1292) Release note: None

There is an pre-existing issue for allowing structured skipping via releases.yaml [1]. Should we close it out in light of this change?

[1] https://github.com/cockroachdb/cockroach/issues/129067

RaduBerinde commented:
Ah, nice. I think we can keep that issue open as a cleanup task to use data from releases instead of hardcoding.
